src/iterable-request.ts

```diff
@@ -0,0 +1,266 @@
+      return false;
+    }
+    const promise = this.promises.shift()!;
+    promise.resolve({ done: true, value: undefined });
+    return true;
+  }
+
+  private serveNextPromise(): boolean {
+    if (this.serveRowItem()) {
+      return true;
+    }
+    if (this.serveError()) {
+      return true;
+    }
+    if (this.serveRequestCompletion()) {
+      return true;
+    }
+    return false;
+  }
+
+  private servePromises() {
+    while (true) {
+      if (!this.serveNextPromise()) {
+        break;
+      }
+    }
+  }
+
+  // This promise executor is called synchronously from within Iterator.next().
+  public promiseExecutor = (resolve: iteratorPromiseResolveFunction, reject: iteratorPromiseRejectFunction) => {
+    this.promises.push({ resolve, reject });
+    this.servePromises();
+  };
+
+  // --- Event handlers --------------------------------------------------------
+
+  public completionCallback(error: Error | null | undefined) {
+    this.requestCompleted = true;
+    if (this.terminating) {
+      if (this.terminatorResolve) {
+        this.terminatorResolve();
+      }
+      return;
+    }
+    if (error && !this.error) {
+      this.error = error;
+    }
+    this.servePromises();
+  }
+
+  private columnMetadataEventHandler = (columnMetadata: ColumnMetadata[] | Record<string, ColumnMetadata>) => {
+    this.resultSetNo++;
+    this.columnMetadata = columnMetadata;
+  };
+
+  private rowEventHandler = (row: RowData) => {
+    if (this.requestCompleted || this.error || this.terminating) {
+      return;
+    }
+    if (this.resultSetNo === 0 || !this.columnMetadata) {
+      this.error = new Error('No columnMetadata event received before row event.');
+      this.servePromises();
+      return;
+    }
+    const item: IterableRequestItem = { row, resultSetNo: this.resultSetNo, columnMetadata: this.columnMetadata };
+    this.fifo.push(item);
+    if (this.fifo.length >= this.fifoPauseLevel && !this.requestPaused) {
+      this.request.pause();
+      this.requestPaused = true;
+    }
+    this.servePromises();
+  };
+
+}
+
+// Internal class for the iterator object which is passed to the API client.
+class IterableRequestIterator implements AsyncIterator<IterableRequestItem> {
+
+  private controller: IterableRequestController;
+
+  constructor(controller: IterableRequestController) {
+    this.controller = controller;
+  }
+
+  public next(): Promise<IteratorResult<IterableRequestItem>> {
+    return new Promise<IteratorResult<IterableRequestItem>>(this.controller.promiseExecutor);
+  }
+
+  public async return(value?: any): Promise<any> {
+    await this.controller.terminate();
+    return Promise.resolve({ value, done: true });         // eslint-disable-line @typescript-eslint/return-await
+  }
+
+  public async throw(exception?: any): Promise<any> {
+    await this.controller.terminate();
+    if (exception) {
+      return Promise.reject(exception);                    // eslint-disable-line @typescript-eslint/return-await
```

Comment:
Could you `throw exception` here?

Reply:
Yes, of course. Thanks. I have changed it.